packages/fast-colors/src/component-state-color-palette.ts

```diff
@@ -0,0 +1,152 @@
+        const baseColorLum2: number = rgbToHSL(this.config.baseColor).l;
+        const baseColorLum: number = (baseColorLum1 + baseColorLum2) / 2;
+        const baseColorRefIndex: number = this.matchRelativeLuminanceIndex(
+            baseColorLum,
+            referencePalette
+        );
+        const baseColorPercent: number = baseColorRefIndex / (steps - 1);
+
+        // Find the luminance location for the dark cutoff.
+        const darkRefIndex: number = this.matchRelativeLuminanceIndex(
+            darkLum,
+            referencePalette
+        );
+        const darkPercent: number = darkRefIndex / (steps - 1);
+
+        // Issue https://github.com/microsoft/fast-dna/issues/1904
+        // Creating a color from H, S, and a known L value is not the inverse of getting the relative
+        // luminace as above. Need to derive a relative luminance version of the color to better match on the dark end.
+
+        // Find the dark cutoff and darkest variations of the requested base color.
+        const baseColorHSL: ColorHSL = rgbToHSL(this.config.baseColor);
+        const darkBaseColor: ColorRGBA64 = hslToRGB(
+            ColorHSL.fromObject({
+                h: baseColorHSL.h,
+                s: baseColorHSL.s,
+                l: darkLum,
+            })
+        );
+        const darkestBaseColor: ColorRGBA64 = hslToRGB(
+            ColorHSL.fromObject({
+                h: baseColorHSL.h,
+                s: baseColorHSL.s,
+                l: darkestLum,
+            })
+        );
+
+        // Create the gradient stops, including the base color and anchor colors for the dark end compression.
+        const fullColorScaleStops: ColorScaleStop[] = new Array(5);
+        fullColorScaleStops[0] = {
+            position: 0,
+            color: new ColorRGBA64(1, 1, 1, 1),
+        };
+        fullColorScaleStops[1] = {
+            position: baseColorPercent,
+            color: this.config.baseColor,
+        };
+        fullColorScaleStops[2] = {
+            position: darkPercent,
+            color: darkBaseColor,
+        };
+        fullColorScaleStops[3] = {
+            position: 0.99,
+            color: darkestBaseColor,
+        };
+        fullColorScaleStops[4] = {
+            position: 1,
+            color: new ColorRGBA64(0, 0, 0, 1),
+        };
+        const scale: ColorScale = new ColorScale(fullColorScaleStops);
+
+        // Create the palette.
+        this.palette = new Array(steps);
+        for (let i: number = 0; i < steps; i++) {
+            const c: ColorRGBA64 = scale.getColor(
+                i / (steps - 1),
+                ColorInterpolationSpace.RGB
+            );
+            this.palette[i] = c;
+        }
+    }
+
+    private matchRelativeLuminanceIndex(input: number, reference: ColorRGBA64[]): number {
+        let bestFitValue: number = Number.MAX_VALUE;
+        let bestFitIndex: number = 0;
+        for (let i: number = 0; i < reference.length; i++) {
```

Comment:
As a perf improvement, you can instead do:
`let i: number = 0, referenceLength: number = reference.length: i < referenceLength; i++`

This will store the reference length when this loop is initialized and not look it up every loop.